src/conversions/std/string.rs

```diff
@@ -233,7 +233,152 @@ impl FromPyObject<'_, '_> for char {
+
+    #[cfg(feature = "experimental-inspect")]
+    fn type_output() -> TypeInfo {
+        <String>::type_output()
+    }
+}
+
+/// Allows extracting cstr from Python objects.
+/// Accepts Python `str` objects and converts them to cstr.
+/// Fails if the string contains interior nul bytes.
+impl<'a> FromPyObject<'a, '_> for &'a std::ffi::CStr {
+    type Error = PyErr;
+
+    #[cfg(feature = "experimental-inspect")]
+    const INPUT_TYPE: &'static str = "str";
+
+    fn extract(obj: Borrowed<'a, '_, PyAny>) -> Result<Self, Self::Error> {
+        let py_string = obj.cast::<PyString>()?;
+
+        // Use PyUnicode_AsUTF8AndSize to get the raw bytes with guaranteed NUL termination
+        #[cfg(any(Py_3_10, not(Py_LIMITED_API)))]
+        {
+            let mut size: ffi::Py_ssize_t = 0;
+            let data: *const u8 =
+                unsafe { ffi::PyUnicode_AsUTF8AndSize(py_string.as_ptr(), &mut size).cast() };
+            if data.is_null() {
+                return Err(crate::PyErr::fetch(obj.py()));
+            }
+
+            // Create a slice that includes the NUL terminator
+            let bytes_with_nul = unsafe { std::slice::from_raw_parts(data, (size + 1) as usize) };
+
+            // Check for interior NUL bytes (excluding the final NUL)
+            for &byte in &bytes_with_nul[..size as usize] {
+                if byte == 0 {
+                    return Err(PyValueError::new_err("string contains interior NUL bytes"));
+                }
+            }
+
+            // Now create the CStr from the bytes with NUL
+            Ok(unsafe { std::ffi::CStr::from_ptr(data as *const i8) })
+        }
+
+        #[cfg(not(any(Py_3_10, not(Py_LIMITED_API))))]
+        {
+            // Fallback for older Python versions
+            use std::ffi::FromBytesWithNulError;
+            let cow = py_string.to_cow()?;
+            let bytes = cow.as_bytes();
+
+            // Try to create a CStr from the bytes
+            // This will fail if there are interior NUL bytes
+            std::ffi::CStr::from_bytes_with_nul(bytes).map_err(|e| match e {
+                FromBytesWithNulError::InteriorNul { .. } => {
+                    PyValueError::new_err("string contains interior NUL bytes")
+                }
+                FromBytesWithNulError::NotNulTerminated => {
+                    PyValueError::new_err("string is not NUL-terminated")
+                }
+            })
+        }
+    }
+
+    #[cfg(feature = "experimental-inspect")]
+    fn type_input() -> TypeInfo {
+        <String as FromPyObject>::type_input()
+    }
+}
+
+/// Allows extracting CString from Python objects.
+/// Accepts Python `str` objects and converts them to CString.
+/// Fails if the string contains NUL bytes.
+impl FromPyObject<'_, '_> for std::ffi::CString {
+    type Error = PyErr;
+
+    #[cfg(feature = "experimental-inspect")]
+    const INPUT_TYPE: &'static str = "str";
+
+    fn extract(obj: Borrowed<'_, '_, PyAny>) -> Result<Self, Self::Error> {
```

Comment:
This should make use of the `CStr` impl above and just call `to_owned` on the result.